### ISSUE TITLE
Consider enabling -O3 optimizations for Release builds

### ISSUE THREAD

**Issue by vegerot** [neovim/neovim#23008]
### Problem

The issue that inspired using `-O2` for Release builds ( #1410 ) is now nearly a decade old (5d97f3f).

Since then, compilers have gotten more reliable `-O3` output.  Should we reconsider using `-O3` in release builds?

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index bfc8a2283..a4b08d01f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -114,12 +114,6 @@ set(NVIM_API_LEVEL 11)        # Bump this after any API change.
 set(NVIM_API_LEVEL_COMPAT 0)  # Adjust this after a _breaking_ API change.
 set(NVIM_API_PRERELEASE false)
 
-# Default to -O2 on release builds.
-if(CMAKE_C_FLAGS_RELEASE MATCHES "-O3")
-  message(STATUS "Replacing -O3 in CMAKE_C_FLAGS_RELEASE with -O2")
-  string(REPLACE "-O3" "-O2" CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE}")
-endif()
-
 # Build-type: RelWithDebInfo
 
 # /Og means something different in MSVC
```

Factors to consider:
1. Reliability.  Does introducing `-O3` cause any bugs?
2. Performance.  Will this make anything slower?  Specifically, is #1410 reintroduced?
3. Build times for Release builds will take longer

Some preliminary measurements:
1. the same tests (with `make test`) pass with this patch
2. Neovim has weak benchmarks, but nonetheless `make benchmark` showed no statistically significant difference
3. I don't think build times for Release are a big concern.  Surprisingly the difference in build times were not statistically significant.  I must have done something wrong

```sh
# with -O2
$ hyperfine --prepare="make deps" --warmup=1 "make -j CMAKE_BUILD_TYPE=Release && make clean"
Benchmark 1: make -j CMAKE_BUILD_TYPE=Release && make clean
  Time (mean ± σ):     63.856 s ±  6.216 s    [User: 2.111 s, System: 1.915 s]
  Range (min … max):   57.523 s … 78.254 s    10 runs
```

```sh
# with -O3
$ hyperfine --prepare="make deps" --warmup=1 "make -j CMAKE_BUILD_TYPE=Release && make clean"
Benchmark 1: make -j CMAKE_BUILD_TYPE=Release && make clean
  Time (mean ± σ):     64.284 s ±  3.792 s    [User: 2.057 s, System: 1.823 s]
  Range (min … max):   59.825 s … 71.872 s    10 runs
```

**Comment by clason** [comment]
Bigger question: Does setting `-O3` have any _advantage_? I don't believe the differences to `-O2` on most compilers are particularly relevant to Neovim.

**Comment by dundargoc** [comment]
I think we should "enable" (not actively disable O3). It will result in less cmake code and we will see some evidence of the claim that -O3 is horribly buggy.

**Comment by clason** [comment]
Removing CMake code is indeed an advantage.

**Comment by vegerot** [comment]
> Bigger question: Does setting `-O3` have any _advantage_? I don't believe the differences to `-O2` on most compilers are particularly relevant to Neovim.

the obvious answer is "measure it".  I did run the benchmarks but they're _very_ sparse.  Do you know of any more comprehensive benchmarks for Neovim?

**Comment by ilan-schemoul** [comment]
I did run some benchmark and could not find statically significant differences
I run this script
```bash
for d in nvim*/ ; do
  hyperfine -r 100 "$d/bin/nvim -c 'call feedkeys(\"10j ciw test\") | q' do_mounts.c" -w 20
done
```
do_mounts.c is a file from linux project (>500 lines)

# 03 (Release mode) march=native

Time (mean ± σ): 241.7 ms ± 13.2 ms [User: 270.6 ms, System: 55.2 ms]

# O3 (Release mode)

Time (mean ± σ): 239.3 ms ± 13.8 ms [User: 263.2 ms, System: 56.3 ms]

# O2

Time (mean ± σ): 240.4 ms ± 14.6 ms [User: 267.5 ms, System: 55.0 ms]

# RelWIthDebInfo

Time (mean ± σ): 248.2 ms ± 15.3 ms [User: 276.1 ms, System: 57.9 ms] 

> Bigger question: Does setting `-O3` have any _advantage_? I don't believe the differences to `-O2` on most compilers are particularly relevant to Neovim.

**Comment by justinmk** [comment]
> I did run some benchmark and could not find statically significant differences

Not surprising. I assume we enabled O3 simply "because we can". If it causes any issues we will just revert.

**Comment by vegerot** [comment]
> because we can

Agreed.  jooc @ilan-schemoul what do you get for `-O1` and `-O0`?

**Comment by dundargoc** [comment]
> I assume we enabled O3 simply "because we can".

We *disabled* O3 in 5d97f3fc2c for "reasons". I reverted that to remove the workaround. The real reason was to have less cmake code and for people to forget about O3 as the conclusion inevitably always end in "I guess it doesn't really matter lmao".

> If it causes any issues we will just revert.

That was basically my stance. If anyone actually finds a real bug and not just a snippet of Linus Torvalds raging against O3 as "evidence" as to why we shouldn't use it then that would be good enough for me.

**Comment by ilan-schemoul** [comment]
> > because we can
> 
> Agreed. jooc @ilan-schemoul what do you get for `-O1` and `-O0`?

Ok. I did it again but this time I used --clean so my plugins don't interfere with the results (and don't slow down the process too much). Note that I check with "diff" command that the binaries were all indeed different from one another.

Benchmark 1: nvim-O0//bin/nvim --clean -c 'call feedkeys("10j ciw test") | q' do_mounts.c
  Time (mean ± σ):      10.2 ms ±   0.5 ms    [User: 7.0 ms, System: 2.7 ms]
  Range (min … max):     9.2 ms …  12.9 ms    800 runs

Benchmark 1: nvim-O1//bin/nvim --clean -c 'call feedkeys("10j ciw test") | q' do_mounts.c
  Time (mean ± σ):       9.5 ms ±   0.5 ms    [User: 6.1 ms, System: 2.9 ms]
  Range (min … max):     8.6 ms …  11.4 ms    800 runs

Benchmark 1: nvim-O2//bin/nvim --clean -c 'call feedkeys("10j ciw test") | q' do_mounts.c
  Time (mean ± σ):       9.6 ms ±   0.5 ms    [User: 6.1 ms, System: 3.0 ms]
  Range (min … max):     8.5 ms …  11.7 ms    800 runs

Benchmark 1: nvim-O3-native//bin/nvim --clean -c 'call feedkeys("10j ciw test") | q' do_mounts.c
  Time (mean ± σ):       9.7 ms ±   0.5 ms    [User: 6.4 ms, System: 2.8 ms]
  Range (min … max):     8.7 ms …  12.3 ms    800 runs

Benchmark 1: nvim-O3//bin/nvim --clean -c 'call feedkeys("10j ciw test") | q' do_mounts.c
  Time (mean ± σ):       9.8 ms ±   0.6 ms    [User: 6.4 ms, System: 2.9 ms]
  Range (min … max):     8.7 ms …  13.0 ms    800 runs

Benchmark 1: nvim-relwithdebinfo//bin/nvim --clean -c 'call feedkeys("10j ciw test") | q' do_mounts.c
  Time (mean ± σ):      10.4 ms ±   0.5 ms    [User: 7.3 ms, System: 2.6 ms]
  Range (min … max):     9.3 ms …  12.3 ms    800 runs

I also timed neovim with the arguments as above. It shows that most time is spent on I/O (makes sense for a text editor) and less than 10% is userland CPU time
Executed in  138.56 millis    fish           external
   usr time   11.87 millis  315.00 micros   11.55 millis
   sys time   30.66 millis  221.00 micros   30.44 millis

If I now time without "--clean" so it means neovim gets to load my 70 plugins I get 30% of user CPU time (but I guess it's mostly lua not compiled code so not much to optimize with -O3 there) : 
Executed in  418.60 millis    fish           external
   usr time  131.81 millis  348.00 micros  131.47 millis
   sys time  109.02 millis  245.00 micros  108.78 millis
